pipeline/plugins/operators/custom_operators.py

```diff
@@ -0,0 +1,346 @@
+
+    def execute(self, context):
+        http_conn = BaseHook.get_connection(self.http_conn_id)
+        api = PMH_API(url=http_conn.host)
+        output_file = str(
+            utils.construct_mets_download_location(
+                dc_identifier=self.dc_identifier,
+                base_path=self.base_path,
+                file_dir="mets",
+            )
+        )
+        mets_path = self.base_path / "mets"
+        mets_path.mkdir(parents=True, exist_ok=True)
+        with open(output_file, "wb") as file:
+            api.download_mets(dc_identifier=self.dc_identifier, output_mets_file=file)
+
+
+class SaveAltosOperator(BaseOperator):
+    """
+    Save ALTO files for one binding on local filesystem.
+
+    :param dc_identifier: DC identifier of binding
+    :param base_path: Base path for download location
+    """
+
+    def __init__(self, dc_identifier, base_path, mets_path, **kwargs):
+        super().__init__(**kwargs)
+        self.dc_identifier = dc_identifier
+        self.base_path = base_path
+        self.mets_path = mets_path
+
+    def execute(self, context):
+        path = os.path.join(
+            self.mets_path, f"{utils.binding_id_from_dc(self.dc_identifier)}_METS.xml"
+        )
+        mets = METS(self.dc_identifier, open(path, "rb"))
+        alto_files = mets.files_of_type(ALTOFile)
+        for alto_file in alto_files:
+            output_file = utils.construct_file_download_location(
+                file=alto_file, base_path=self.base_path
+            )
+            output_file.parent.mkdir(parents=True, exist_ok=True)
+            with open(output_file, "wb") as file:
+                alto_file.download(output_file=file)
+
+
+class SaveMetsSFTPOperator(BaseOperator):
+    """
+    Save METS file for one binding on remote filesystem using SSH connection.
+
+    :param http_conn_id: Connection ID of API
+    :param sftp_client: SFTPClient
+    :param dc_identifier: DC identifier of binding
+    :param base_path: Base path for download location
+    """
+
+    def __init__(self, api, sftp_client, dc_identifier, base_path, file_dir, **kwargs):
+        super().__init__(**kwargs)
+        self.api = api
+        self.sftp_client = sftp_client
+        self.dc_identifier = dc_identifier
+        self.base_path = base_path
+        self.file_dir = file_dir
+
+    def execute(self, context):
+        output_file = str(
+            utils.construct_mets_download_location(
+                dc_identifier=self.dc_identifier,
+                base_path=self.base_path,
+                file_dir=self.file_dir,
+            )
+        )
+
+        with self.sftp_client.file(output_file, "w") as file:
+            try:
+                self.api.download_mets(
+                    dc_identifier=self.dc_identifier, output_mets_file=file
+                )
+            except HTTPError as e:
+                print(
+                    f"Download of METS file {self.dc_identifier} failed with args {e.args}"
```

Comment:
I googled around a bit, and it might be that we are better off using `logger` (https://stackoverflow.com/a/41620340 style): that way the log level shows up in the output and we can e.g. use the log level filtering in the web UI. Actually I wasn't able do dig up the `print`ed logs from there at all, but that might be just me not being diginatiivi enough.

This is not something we are in a hurry to change though: we can just as well work on the logging when we do the parallelization refactoring.

Reply:
For sure. Let's add proper logging when we work on the parallelization.